### PR TITLE
Select: Portal select menu to document.body

### PR DESCRIPTION
**Pull request by ashharrison90**
**What this PR does / why we need it**:

- portals the `Select` menu to `document.body`
- `auto` positions the `Select` menu for better placement
- closes the `Select` menu when scrolling anywhere outside of the `Select` menu
- adds a new `portal` z-index value set to higher than the modal
- rewrites `Portal` component as a functional component so it can utilise `useTheme2`
- rewrites `SelectBase` tests as RTL tests
- updates other unit tests
  - `react-select-event` has an extra option to specify where the menu is contained. e.g.
    ```javascript
    await selectEvent.select(selectEl, 'Option 2');
    ```
    becomes
    ```javascript
    await selectEvent.select(selectEl, 'Option 2', { container: document.body });
    ```
    or use the new util method exposed from `@grafana/ui`:
    ```javascript
    import { selectOptionInTest } from '@grafana/ui';
    await selectOptionInTest(selectEl, 'Option 2');
    ```
- updates e2e tests
  - as the menu is no longer within the `Select` container, need to look for options on the body. e.g.
    ```javascript
    e2e.components.TimeZonePicker.container()
      .should('be.visible')
      .within(() => {
        e2e.components.Select.input().should('be.visible').click();
        e2e.components.Select.option().should('be.visible').contains(toTimeZone).click();
      });
    ``` 
    should now be written as 
    ```javascript
    e2e.components.TimeZonePicker.container()
      .should('be.visible')
      .within(() => {
        e2e.components.Select.input().should('be.visible').click();
      });
    e2e.components.Select.option().should('be.visible').contains(toTimeZone).click();
    ```
- have also updated the e2e util method `selectOption` here: https://github.com/grafana/grafana/blob/main/packages/grafana-e2e/src/flows/selectOption.ts to work with the new structure.
- since the menu is portalled to body, wrapping the `Select` in something like `ClickOutsideWrapper` will no longer work. the same functionality can be achieved using `Select`'s `onCloseMenu`/`isOpen` props.

some screenshots:
image
image
image

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by torkelo** [comment]
Awesome job! this is tricky stuff, and we have had max-width / overflow hidden/auto before in the Modal but it issues (esp with selects) made us change that back. There are some other Modals with selects in them (like dashboard SaveAs modal).

**Review by Clarity-89 — APPROVED** [review]
Looks good! We don't have any use cases of Selects in Modal in Enterprise as far as I can see. Regarding the value mappings and modal in general, I think a nice improvement would be if the button actions wouldn't be a pat of scrolling content and always stayed visible during overflow, but this is probably out of scope of this PR.

**Comment by ashharrison90** [comment]
@Clarity-89 +1 on that. i think every major design library out there (bootstrap, material, carbon, etc.) separates the actions from the content for this reason.

**Comment by torkelo** [comment]
An idea would to add a footer prop to Modal that automatically wraps the footer content in a ModalButtonRow, and the footer would live outside the content div

File: packages/grafana-ui/src/components/Select/types.ts [hunk 1]
```diff
@@ -41,6 +41,7 @@ export interface SelectCommonProps<T> {
   minMenuHeight?: number;
   maxVisibleValues?: number;
   menuPlacement?: 'auto' | 'bottom' | 'top';
+  menuPortalTarget?: HTMLElement;
```
[the remark this answers is not in the document]
Comment (tskarhed) [reply packages/grafana-ui/src/components/Select/types.ts:44]: Maybe it is okay in this case

**Comment by ashharrison90** [comment]
> An idea would to add a footer prop to Modal that automatically wraps the footer content in a ModalButtonRow, and the footer would live outside the content div

yep, i don't make raising that/taking it on as an enhancement following this PR 👍

File: packages/grafana-ui/src/components/Select/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashharrison90) [reply packages/grafana-ui/src/components/Select/types.ts:44]: hmm, that's a good point. i'm not sure how we can avoid it though.
we could attach all selects to `document.body` by default? that would have a pretty wide impact though, and is presumably slightly worse for performance.
ideally we'd want something where the select detects it's in a modal and attaches itself to document.body automatically 🤔

[the remark this answers is not in the document]
Comment (tskarhed) [reply packages/grafana-ui/src/components/Select/types.ts:44]: I played around a bit with the portaling etc. last year, and it never ended up good. When this is merged we might have to keep an eye out for Select bugs.

Let me just check a few places before you merge this PR.

**Review by tskarhed — APPROVED** [review]
Okay, let's go! Be on the lookout for weird Select bugs though!

**Comment by ashharrison90** [comment]
@tskarhed @torkelo what do you think about using context to determine whether the `Select` component is in a modal or not? 🤔 

we could export a context object from `SelectBase` - something like `IsInModalContext` - default it to false and wrap the modal like:
```jsx
    <Portal>
      <div className={cx(styles.modal, className)}>
        <IsInModalContext.Provider value={true}>
          <div className={headerClass}>
            {typeof title === 'string' && <DefaultModalHeader {...props} title={title} />}
            {typeof title !== 'string' && title}
            <div className={styles.modalHeaderClose}>
              <IconButton surface="header" name="times" size="xl" onClick={onDismiss} />
            </div>
          </div>
          <div className={cx(styles.modalContent, contentClassName)}>{children}</div>
        </IsInModalContext.Provider>
      </div>
      <div
        className={styles.modalBackdrop}
        onClick={onClickBackdrop || (closeOnBackdropClick ? onDismiss : undefined)}
      />
    </Portal>
```

that way we wouldn't have to expose any new props on `SelectBase` or `ValuePicker`.

**Comment by ashharrison90** [comment]
ok, i've updated this to portal `Select` to `document.body` in all cases and updated the unit tests accordingly. the select menu now also closes when you scroll outside of the select menu.

the only thing left to discuss is what we do about z-index. in this PR the select menu overlays everything, including modals. i can imagine there's cases where that's not desirable.

but when a select is within a modal, it **has** to overlay the modal, else it's broken. so it seems like we'll need to make it configurable in some way. either by exposing a prop, or setting it via context, or something else.

File: public/app/plugins/datasource/influxdb/components/RawInfluxQLEditor.test.tsx [hunk 2]
```diff
@@ -57,7 +57,7 @@ describe('RawInfluxQLEditor', () => {
     const formatSelect = screen.getByLabelText('Format as');
     expect(formatSelect).toBeInTheDocument();
 
-    await select(formatSelect, 'Time series');
+    await select(formatSelect, 'Time series', { container: document.body });
```
Comment (torkelo) [inline public/app/plugins/datasource/influxdb/components/RawInfluxQLEditor.test.tsx:60]: this seems like a big pain to always have to specify in tests

File: packages/grafana-ui/src/themes/default.ts [hunk 3]
```diff
@@ -123,15 +135,7 @@ const theme: GrafanaThemeCommons = {
...
-  zIndex: {
-    navbarFixed: 1000,
-    sidemenu: 1020,
-    dropdown: 1030,
-    typeahead: 1030,
-    tooltip: 1040,
-    modalBackdrop: 1050,
-    modal: 1060,
-  },
+  zIndex: appZIndexes,
```
Comment (torkelo) [inline packages/grafana-ui/src/themes/default.ts:138]: this is the old theme, new theme we introduced in v8 is here https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/themes/zIndex.ts  ,  think we missed making this use properties from the new theme

File: packages/grafana-ui/src/components/Portal/Portal.tsx [hunk 4]
```diff
@@ -1,5 +1,6 @@
 ﻿import React, { PureComponent } from 'react';
 import ReactDOM from 'react-dom';
+import { appZIndexes } from '../../themes/default';
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Portal/Portal.tsx:3]: Feels cleaner to rewrite this a a functional component and use useTheme2 to access the theme object (that has zIndex)

**Comment by torkelo** [comment]
Awesome job on this one ash! think it's working well, the hide on scroll is also working. 

A bit afraid of this change, but it does simplify things to always portal the select. No longer any restrictions on usage or special things we need to remember when placing select inside CustomScrollbars or Modal). The impact to testing of Select is the thing that makes me a bit worried (I know Selects are used heavily in e2e tests in plugins as well). Can we introduce a util function we can reuse to share the logic of selecting a menu item? Won't solve the e2e / breaking issue but looks impossible to avoid that

File: public/app/features/alerting/unified/AmRoutes.test.tsx [hunk 5]
```diff
@@ -343,13 +344,13 @@ describe('AmRoutes', () => {
...
+  await selectEvent.select(selectElement, optionText, { container: document.body });
 };
 
 const updateTiming = async (selectElement: HTMLElement, value: string, timeUnit: string): Promise<void> => {
   const inputs = byRole('textbox').queryAll(selectElement);
   expect(inputs).toHaveLength(2);
   await userEvent.type(inputs[0], value);
   userEvent.click(inputs[1]);
-  userEvent.click(byText(timeUnit).get(selectElement));
+  await selectEvent.select(selectElement, timeUnit, { container: document.body });
```
[the remark this answers is not in the document]
Comment (ashharrison90) [reply public/app/features/alerting/unified/AmRoutes.test.tsx:355]: i could be wrong, but i believe this is a menu element it's clicking

**Comment by ashharrison90** [comment]
this has ended up being a much more significant change than originally intended. 

i've updated the description of the PR and removed the modal changes so this will focus solely on the `Select` portalling. i'll deliver the modal changes as part of a separate (much smaller) PR.

i've also moved the milestone to 8.1 and marked it as a breaking change. strictly speaking there's no breaking api changes, but it's a change in behaviour and definitely a breaking change for anyone relying on the old behaviour in unit/e2e tests. 

@torkelo think i've addressed your comments apart from the util function. i'll start looking into some sort of util function for the e2e side

**Comment by ashharrison90** [comment]
i believe there's already an e2e method here: https://github.com/grafana/grafana/blob/ash/36317/packages/grafana-e2e/src/flows/selectOption.ts#L11

i've updated it to search for options outside of the container now 👍

File: public/app/plugins/datasource/influxdb/components/RawInfluxQLEditor.test.tsx (hunk 2 above)
Comment (mckn) [reply public/app/plugins/datasource/influxdb/components/RawInfluxQLEditor.test.tsx:60]: Maybe add a `select` function to grafana-ui that basically just does this by default?

**Review by mckn — APPROVED** [review]
Great work on this @ashharrison90! Looks really good and seems to work in the scenarios that I have tested it.

**Comment by mckn** [comment]
Love that you included instructions on how to change the tests in this PR! Will help people that might run into this issue! 🙌🏻

**Review by torkelo — APPROVED** [review]
Awesome job on this ash!

only thing that can be improved would be a test helper function to select options so you not have to always pass in the container argument, should make it easier for to change this in the future without updating all tests (I hope we do no need to!!)

**Comment by ashharrison90** [comment]
unit test util method created. once everything goes green i'll merge 🤞

File: packages/grafana-ui/src/components/Select/SelectBase.tsx [hunk 6]
```diff
@@ -329,19 +334,16 @@ export function SelectBase<T>({
         }}
         styles={{
           ...resetSelectStyles(),
-          menuPortal: ({ position, width }: any) => ({
-            position,
-            width,
-            zIndex: theme.zIndex.dropdown,
+          menuPortal: (base: any) => ({
+            ...base,
+            zIndex: theme.zIndex.portal,
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Select/SelectBase.tsx:339]: was that it? :)

Comment (ashharrison90) [reply packages/grafana-ui/src/components/Select/SelectBase.tsx:339]: oooo maybe, forgot about that change 😂

**Pull request merged**